### PR TITLE
Feature/add test

### PR DESCRIPTION
**Pull request by Buber11**

### CODE REVIEW

**Comment by MateuszBlach** [comment]
Please change comments im code to english

**Comment by Buber11** [comment]
done

File: src/common/test/Test.hpp [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+#include "interfaces/ISieve.hpp"
+
+class Test {
+public:
+    void runTest(long max, std::shared_ptr<ISieve> sieve, std::ofstream& outFile);
+private:
+    static std::shared_ptr<Timer> timer;
+};
+
+#endif
```
Comment (Jocho14) [inline src/common/test/Test.hpp]: #endif  // TEST_HPP

File: src/common/test/Test.cpp
Comment (Jocho14) [inline src/common/test/Test.cpp:1]: move outside of common to src/test

File: src/common/test/Test.hpp
Comment (Jocho14) [inline src/common/test/Test.hpp:1]: move outside of common to src/test

File: src/gpu/GpuSieve.hpp [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    GpuSieve();
+    ~GpuSieve() override;
+
+    void setMaxLimit(unsigned int maxLimit) override;
+    void computePrimes() override;
+    const std::vector<unsigned int>& getPrimes() const override;
+    void collectPrimes() override;
+
+private:
+    unsigned int maxLimit;
```
Comment (Jocho14) [inline src/gpu/GpuSieve.hpp]: add underscores suffixes for any private member, in this case: unsigned int maxLimit_;

File: src/common/test/Test.hpp [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+#include <fstream>
+
+#include "common/timer/Timer.hpp"
+#include "interfaces/ISieve.hpp"
+
+class Test {
+public:
+    void runTest(long max, std::shared_ptr<ISieve> sieve, std::ofstream& outFile);
+private:
+    static std::shared_ptr<Timer> timer;
```
Comment (Jocho14) [inline src/common/test/Test.hpp]: remove static and add underscore suffix: timer_;

File: src/common/test/Test.cpp [hunk 4]
```diff
@@ -0,0 +1,21 @@
+#include "test.hpp"
+
+
+std::shared_ptr<Timer> Test::timer = std::make_shared<Timer>();
```
Comment (Jocho14) [inline src/common/test/Test.cpp]: remove this and initialize the timer in the constructor

File: src/gpu/GpuSieve.hpp [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    bool* isPrimeListHost;
```
Comment (Jocho14) [inline src/gpu/GpuSieve.hpp]: try to use unique pointer here
std::unique_ptr<bool> isPrimeListHost_;

File: src/gpu/GpuSieve.cu [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    }
+}
+
+// Function to launch GPU sieve with prepared primes
+void GpuSieve::gpuSieve(uint64_t maxNumber, const std::vector<uint64_t>& prepedPrimes) {
+    bool* isPrimeListDevice = nullptr;
+    uint64_t* prepedPrimesDevice = nullptr;
+
+    // Allocate and initialize memory on host and device
+    isPrimeListHost = (bool*)malloc(maxNumber * sizeof(bool));
```
Comment (Jocho14) [inline src/gpu/GpuSieve.cu]: isPrimeListHost_ = std::make_unique<bool[]>(maxNumber);

File: src/gpu/GpuSieve.cu [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    std::memset(isPrimeListHost, true, maxNumber * sizeof(bool));
```
Comment (Jocho14) [inline src/gpu/GpuSieve.cu]: as you are using unique ptr change this to:
std::memset(isPrimeListHost_.get(), true, maxNumber * sizeof(bool));

File: src/gpu/GpuSieve.cu [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    cudaMalloc(&isPrimeListDevice, maxNumber * sizeof(bool));
+    cudaMemcpy(isPrimeListDevice, isPrimeListHost, maxNumber * sizeof(bool), cudaMemcpyHostToDevice);
+
+    // Prepare and copy the list of initial primes to device
+    uint64_t prepedPrimesSize = prepedPrimes.size();
+    uint64_t* prepedPrimesHost = (uint64_t*)malloc(prepedPrimesSize * sizeof(uint64_t));
```
Comment (Jocho14) [inline src/gpu/GpuSieve.cu]: use unique pointer with "auto" keyword:
auto prepedPrimesHost = std::make_unique<uint64_t[]>(prepedPrimesSize);

File: src/gpu/GpuSieve.cu [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+// Constructor: Initialize GpuSieve object with no allocated memory on the host
+GpuSieve::GpuSieve() : maxLimit(0), isPrimeListHost(nullptr) {}
+
+// Destructor: Free host memory if allocated
+GpuSieve::~GpuSieve() {
+    // Free host memory
+    if (isPrimeListHost) {
+        free(isPrimeListHost);
+        isPrimeListHost = nullptr;
+    }
```
Comment (Jocho14) [inline src/gpu/GpuSieve.cu]: leave the constructor empty as unique pointer handles memory dealocation automatically

File: src/gpu/GpuSieve.cu [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    if (numberOfBlocks * BLOCK_SIZE > MAX_STRIDE) {
+        numberOfBlocks = MAX_STRIDE / BLOCK_SIZE;
+    }
+
+    // Launch the kernel
+    gpuSieveKernel << <numberOfBlocks, BLOCK_SIZE >> > (maxNumber, isPrimeListDevice, prepedPrimesDevice, prepedPrimesSize);
+    cudaDeviceSynchronize();
+
+    // Copy results back to the host
+    cudaMemcpy(isPrimeListHost, isPrimeListDevice, maxNumber * sizeof(bool), cudaMemcpyDeviceToHost);
```
Comment (Jocho14) [inline src/gpu/GpuSieve.cu]: cudaMemcpy(isPrimeListHost_.get(), isPrimeListDevice, maxNumber * sizeof(bool), cudaMemcpyDeviceToHost);

**Review by Jocho14 — COMMENTED** [review]
please resolve the comments

File: src/gpu/GpuSieve.cu [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+// Constructor: Initialize GpuSieve object with no allocated memory on the host
+GpuSieve::GpuSieve() : maxLimit(0), isPrimeListHost_(nullptr) {}
+
+// Destructor: Free host memory if allocated
+GpuSieve::~GpuSieve() {}
+
+// Set the maximum limit for prime computation
+void GpuSieve::setMaxLimit(unsigned int maxLimit) {
+    this->maxLimit = maxLimit;
```
Comment (Jocho14) [inline src/gpu/GpuSieve.cu]: maxLimit_ = maxLimit;

File: src/gpu/GpuSieve.cu [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+    // Prepare initial list of small primes on CPU using a simple sieve
+    std::vector<uint64_t> prepedPrimes = sieveCpuPrep(std::sqrt(maxLimit));
+
+    // Launch GPU sieve to find all primes up to maxLimit
+    gpuSieve(maxLimit, prepedPrimes);
+}
+
+// Retrieve the list of primes found
+const std::vector<unsigned int>& GpuSieve::getPrimes() const {
+    return primes;
```
Comment (Jocho14) [inline src/gpu/GpuSieve.cu]: return primes_;

File: src/gpu/GpuSieve.cu [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+// Set the maximum limit for prime computation
+void GpuSieve::setMaxLimit(unsigned int maxLimit) {
+    this->maxLimit = maxLimit;
+}
+
+// Main function to compute primes up to maxLimit
+void GpuSieve::computePrimes() {
+    // Prepare initial list of small primes on CPU using a simple sieve
+    std::vector<uint64_t> prepedPrimes = sieveCpuPrep(std::sqrt(maxLimit));
```
Comment (Jocho14) [inline src/gpu/GpuSieve.cu]: std::vector<uint64_t> prepedPrimes = sieveCpuPrep(std::sqrt(maxLimit_));

File: src/gpu/GpuSieve.cu [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+const std::vector<unsigned int>& GpuSieve::getPrimes() const {
+    return primes;
+}
+
+// Collect primes from the result list on the host
+void GpuSieve::collectPrimes() {
+    // Clear any existing primes in the list
+    primes.clear();
+    // Traverse isPrimeListHost to collect primes up to maxLimit
+    for (uint64_t i = 2; i <= maxLimit; i++) {
```
Comment (Jocho14) [inline src/gpu/GpuSieve.cu]: for (uint64_t i = 2; i <= maxLimit_; i++) {

**Pull request merged**